File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.py
class Solution:
    def earliestFinishTime(
        self,
        landStartTime: List[int],
        landDuration: List[int],
        waterStartTime: List[int],
        waterDuration: List[int],
    ) -> int:
        MAX = 10**6
        m, n = len(landStartTime), len(waterStartTime)
        land = water = minLand = minWater = MAX

        for s, d in zip(landStartTime, landDuration):
            land = min(s + d, land)

        for s, d in zip(waterStartTime, waterDuration):
            water = min(s + d, water)

        for s, d in zip(landStartTime, landDuration):
            minWater = min(minWater, max(water, s) + d)

        for s, d in zip(waterStartTime, waterDuration):
            minLand = min(minLand, max(land, s) + d)

        return min(minLand, minWater)

File: 3967-earliest-finish-time-for-land-and-water-rides-ii/earliest-finish-time-for-land-and-water-rides-ii.cpp
class Solution {
public:
    int earliestFinishTime(vector<int>& landStartTime,
                           vector<int>& landDuration,
                           vector<int>& waterStartTime,
                           vector<int>& waterDuration) {
        int water = INT_MAX;
        int land = INT_MAX;
        int waterAns = INT_MAX;
        int landAns = INT_MAX;

        int m = landStartTime.size();
        int n = waterStartTime.size();

        for (int i = 0; i < m; ++i) {
            land = min(land, landStartTime[i] + landDuration[i]);
        }

        for (int i = 0; i < n; ++i) {
            water = min(water, waterStartTime[i] + waterDuration[i]);
        }

        for (int i = 0; i < m; ++i) {
            waterAns =
                min(waterAns, max(water, landStartTime[i]) + landDuration[i]);
        }

        for (int i = 0; i < n; ++i) {
            landAns =
                min(landAns, max(land, waterStartTime[i]) + waterDuration[i]);
        }

        return min(waterAns, landAns);
    }
};